packages/flutter/lib/src/cupertino/segmented_control.dart

```diff
@@ -0,0 +1,568 @@
+              child: Semantics(
+                inMutuallyExclusiveGroup: true,
+                selected: _selectedValue != null ? _selectedValue == currentKey : false,
+                child: child,
+              ),
+            )),
+      );
+      _gestureChildren.add(child);
+    }
+
+    final Widget box = new _SegmentedControlRenderWidget<T>(
+      children: _gestureChildren,
+      selectedValue: _selectedValue,
+      pressedValue: _pressedValue,
+      keys: widget.children.keys.toList(),
+    );
+
+    return Padding(
+      padding: _kHorizontalItemPadding.resolve(Directionality.of(context)),
+      child: UnconstrainedBox(
+        constrainedAxis: Axis.horizontal,
+        child: box,
+      ),
+    );
+  }
+}
+
+class _SegmentedControlRenderWidget<T> extends MultiChildRenderObjectWidget {
+  _SegmentedControlRenderWidget({
+    Key key,
+    List<Widget> children = const <Widget>[],
+    @required this.selectedValue,
+    @required this.pressedValue,
+    @required this.keys,
+  }) : super(
+          key: key,
+          children: children,
+        );
+
+  final T selectedValue;
+  final T pressedValue;
+  final List<T> keys;
+
+  @override
+  RenderObject createRenderObject(BuildContext context) {
+    return new _RenderSegmentedControl<T>(
+      textDirection: Directionality.of(context),
+      selectedValue: selectedValue,
+      pressedValue: pressedValue,
+      keys: keys,
+    );
+  }
+
+  @override
+  void updateRenderObject(BuildContext context, _RenderSegmentedControl<T> renderObject) {
+    renderObject
+      ..textDirection = Directionality.of(context)
+      ..selectedValue = selectedValue
+      ..pressedValue = pressedValue;
+  }
+}
+
+class _SegmentedControlContainerBoxParentData extends ContainerBoxParentData<RenderBox> {}
+
+class _RenderSegmentedControl<T> extends RenderBox
+    with ContainerRenderObjectMixin<RenderBox, ContainerBoxParentData<RenderBox>> {
+  _RenderSegmentedControl({
+    List<RenderBox> children,
+    @required T selectedValue,
+    @required T pressedValue,
+    @required TextDirection textDirection,
+    @required this.keys,
+  })  : _textDirection = textDirection,
+        _selectedValue = selectedValue,
+        _pressedValue = pressedValue {
+    addAll(children);
+  }
+
+  T get selectedValue => _selectedValue;
+  T _selectedValue;
+  set selectedValue(T value) {
+    if (_selectedValue == value) {
+      return;
+    }
+    _selectedValue = value;
+    markNeedsPaint();
+  }
+
+  T get pressedValue => _pressedValue;
+  T _pressedValue;
+  set pressedValue(T value) {
+    if (_pressedValue == value) {
+      return;
+    }
+    _pressedValue = value;
+    markNeedsPaint();
+  }
+
+  TextDirection get textDirection => _textDirection;
+  TextDirection _textDirection;
+  set textDirection(TextDirection value) {
+    if (_textDirection == value) {
+      return;
+    }
+    _textDirection = value;
+    markNeedsLayout();
+  }
+
+  List<T> keys;
```

Comment:
what if the keys change? do you not need to repaint?

Reply:
(this is a problem that goes away if you precompute the indices in the build method)